API change, AudioConverter interface and BitAudioCoverter class

Riyria-was-taken opened the pull request:
- Changed library API to meet the OOP paradigm
- Added IAudioConverter interface and derived class BitAudioConverter - to encapsulate data to audio encoding and decoding
- Implemented data to audio encoding in BitAudioConverter
- Changed files which required that after listed modifications
- Did style fixes

Riyria-was-taken committed:
WIP: data to send encoding

Riyria-was-taken committed:
Changed library API to meet the OOP paradigm, added IAudioConverter interface and derived class - BitAudioConverter to encapsulate data to audioencoding and decoding. Also, changed files which required that after listed modifications and did style fixes

Riyria-was-taken committed:
Style fix

Lorak-mmk commented on libecho/CMakeLists.txt:
```diff
@@ -15,22 +15,24 @@ endif ()
+        ${SRC_DIR}/AudioStream.h
+	${SRC_DIR}/IAudioConverter.h
+	${SRC_DIR}/BitAudioConverter.h)
 
```
Wrong indentation

Lorak-mmk commented on libecho/src/BitAudioConverter.h:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    T audio_wave(int frequency, int time) {
+        return sin((double)frequency * time / this->outputFormat.sampleRate()) *
```
sin -> std::sin

Lorak-mmk replied:
(double)frequency - I think clang-tidy doesn't check headers, I probably misconfigured it. But we should avoid c-style casts, maybe it would be better to change it to static_cast<double>(frequency)

Lorak-mmk commented on libecho/src/Echo.cc:
```diff
@@ -0,0 +1,56 @@
+}
+
+void Echo::initEcho(int a_argc, char** a_argv) {
+    static QTInitializer init(a_argc, a_argv);
```
You said yourself it would be better to use {} that () :p

Riyria-was-taken replied:
I copied it from your code XD

Lorak-mmk replied:
Nope

Lorak-mmk commented on libecho/src/IAudioConverter.h:
```diff
@@ -0,0 +1,36 @@
+        : inputFormat(inputFormat), outputFormat(outputFormat), windowLength(windowLength){};
+
+    std::vector<char> encode(std::vector<uint8_t> data) {
+        std::vector<T> result;
```
Maybe try to estimate size of result and reserve space in vector? We are dealing with real-time audio coding, we should care about performance

Lorak-mmk commented on libecho/src/Echo.cc:
```diff
@@ -0,0 +1,56 @@
+void Echo::send(const std::vector<uint8_t>& buffer) {
+    auto encoded = converter->encode(buffer);
+    const size_t atOnce = 260;  // (bitrate / notify_interval) + eps; - TODO: wyjaśnić
+    output->enqueueData(encoded.data(), std::min(encoded.size(), atOnce));
```
Maybe a newline before this line? And generally, newlines across code to separate code blocks into logical parts

Lorak-mmk commented on libecho/src/Echo.cc:
```diff
@@ -0,0 +1,56 @@
+
+std::vector<uint8_t> Echo::receive() {
+    auto info = input->getStreamInfo();
+
```
This newline seems unnecessary

Lorak-mmk commented on libecho/src/Echo.cc:
```diff
@@ -0,0 +1,56 @@
+    const size_t atOnce = info.periodSize;
+    char* array = new char[atOnce];
+    int idx = 0;
+    while (input->getStreamStatus().second < 2000000) {
```
But newline here would be great

Lorak-mmk commented on libecho/src/Echo.cc:
```diff
@@ -0,0 +1,56 @@
+    input = std::make_unique<AudioInput>(inputFormat);
+    auto outputFormat = AudioFormatFactory::getDefaultOutputFormat();
+    output = std::make_unique<AudioOutput>(outputFormat);
+    converter = std::make_unique<BitAudioConverter<SampleType>>(inputFormat, outputFormat, windowLength, loFreq, hiFreq);
```
Maybe some newlines? before both auto and before last line?

Lorak-mmk commented on libecho/src/IAudioConverter.h:
```diff
@@ -0,0 +1,36 @@
+            auto encoded_byte = encode_byte(byte);
+            result.insert(result.end(), encoded_byte.begin(), encoded_byte.end());
+        }
+        return std::vector<char>(result.data(), result.data() + result.size());
```
newline before return?

PawelA commented on libecho/include/Echo.h:
```diff
@@ -0,0 +1,50 @@
+
+private:
+    using SampleType = int16_t;
+    std::unique_ptr<AudioInput> input;
+    std::unique_ptr<AudioOutput> output;
+    std::unique_ptr<IAudioConverter<SampleType>> converter;
```
Why the pointers? What's wrong with AudioInput input; etc.?

Lorak-mmk replied:
Because if those ar values, we can't initialize them ourselves in constructor body (not initializer list), C++ tries to call empty constructor (which doesn't exists). If you know how to fix this, please tell

Lorak-mmk commented on libecho/src/IAudioConverter.h:
```diff
@@ -0,0 +1,36 @@
+    IAudioConverter(QAudioFormat inputFormat, QAudioFormat outputFormat, double windowLength)
+        : inputFormat(inputFormat), outputFormat(outputFormat), windowLength(windowLength){};
+
+    std::vector<char> encode(std::vector<uint8_t> data) {
```
Maybe this method should be marked virtual? We may want to encode data differently. Same goes for decode

Riyria-was-taken replied:
Actually, I think it will be like that for nearly every data encoding and I implemented it here to avoid pasting same code everywhere. In case we want to implement it differently in some derived class we can always override method.

Riyria-was-taken replied:
When it comes to decode method, it is up to @PawelA to decide how he wants to deal with that as decoding is his part. I just declared it here like that to be able to compile code.

Lorak-mmk replied:
We can override method if it is marked as virtual, which is the very reason I asked for it.

Riyria-was-taken replied:
Ah xd I misread that it should be purely virtual.

Lorak-mmk reviewed (changes requested):
Please remember about commenting your code and documenting encoding algorithm.

a-piasta commented on libecho/src/IAudioConverter.h:
```diff
@@ -0,0 +1,36 @@
+            auto encoded_byte = encode_byte(byte);
+            result.insert(result.end(), encoded_byte.begin(), encoded_byte.end());
+        }
+        return std::vector<char>(result.data(), result.data() + result.size());
```
niezbyt mi się podoba pomysł, żeby wrzucać wszystko do vectora, a następnie tworzyć na jego podstawie nowy wyłącznie po to, aby przekonwertować typy

Riyria-was-taken replied:
To jest tymczasowe rozwiązanie, będę pracować nad sensownym podejściem do konwersji typów w tym miejscu. Plusem takiego rozwiązania jest to, że nie mamy reinterpret_castów.

a-piasta replied:
reinterpret_cast na char* akurat ma poprawnie zdefiniowane w standardzie zachowanie, więc nie widzę powodu, żeby go na siłę unikać (szczególnie, że w tym przypadku program prawdopodobnie mógłby działać szybciej)

Riyria-was-taken replied:
Rozumiem, wezmę to pod uwagę i rozwiążę w przyszłym PR kiedy to będę się zajmować ładną konwersją typów. Nie będę tego zmieniać na ten moment, ponieważ będzie to wymagało niepotrzebnej, tymczasowej zmiany interfejsu.

a-piasta commented on libecho/src/Echo.cc:
```diff
@@ -0,0 +1,56 @@
+    auto info = input->getStreamInfo();
+
+    const size_t atOnce = info.periodSize;
+    char* array = new char[atOnce];
```
memleak?
https://stackoverflow.com/questions/21377360/proper-way-to-create-unique-ptr-that-holds-an-allocated-array

a-piasta commented on libecho/src/BitAudioConverter.h:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    T audio_wave(int frequency, int time) {
+        return sin((double)frequency * time / this->outputFormat.sampleRate()) *
```
czym jest zwracany typ T? jeżeli zakładamy, że to jeden z podstawowych typów zmiennoprzecinkowych (float, double, long double), to raczej powinniśmy frequency castować konkretnie na T, a nie na double, aby uniknąć wielokrotnego castowania

Lorak-mmk replied:
Typ zwracany to char/short. A ten cast na doubla to chyba po to żeby do sin przekazac doubla, nie ma nic wspolnego z typem zwracanym

Riyria-was-taken replied:
W zależności od tego, jak będzie ustawione SampleSize oraz SampleType w AudioFormatFactory będą to różne typy, nie zawsze zmiennoprzecinkowe (na ten moment jest to int16_t). W związku z tym nie możemy castować po prostu na T.

a-piasta reviewed (changes requested):
kocham anime

Riyria-was-taken committed:
IAudioConverter efficiency fix, IAudioConverter method's headers update, clang-format.sh fix

- Added encryptedByteSize class member to IAudioConverter. We can improve performance in encode method by using this member to reserve needed vector space.
- Updated encode, decode methods in IAudioConverter to be virtual.
- Fixed clang-format.sh to format *.cc files as well.
- Resolved PR style comments.

Lorak-mmk reviewed (approved):
Let's merge it, but we need to document this code and change send() method in Echo.cc before sprint ends.
